config/monitoring/metrics/opencensus/100-ocagent.yaml

```diff
@@ -0,0 +1,104 @@
+# Copyright 2018 The Knative Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Based on https://github.com/census-instrumentation/opencensus-service/raw/master/example/k8s.yaml
+
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: opencensus-agent-config
+  namespace: knative-monitoring
+  labels:
+    serving.knative.dev/release: devel
+data:
+  opencensus-agent-config: |
+    receivers:
+      opencensus:
+        address: ":55678"
+    #   jaeger: {}
+    #   zipkin: {}
+    exporters: {}
+    # stackdriver:
+    #   project: "my-gcp-project"  # Optional, defaults to node project
+    #   metric_prefix: "your-metric-prefix"
+    #   enable_metrics: true
+    #   enable_tracing: true
```

Comment:
Looking at https://github.com/census-instrumentation/opencensus-service/blob/master/exporter/stackdriverexporter/stackdriver.go, these four are the whole available config fields for Stackdriver exporter. Per this [line](https://github.com/census-instrumentation/opencensus-service/blob/fb16513301ba831e33020eccb21198292402b9d3/exporter/stackdriverexporter/stackdriver.go#L91), I highly suspect opencensus agent only supports stackdriver custom metrics using `global` resource type. It doesn't support stackdriver built-in resource type, which we are currently sending metrics to. For example [`knative_revision`](https://cloud.google.com/monitoring/api/metrics_knative). Migrating to opencensus agent, which is in maintain mode, will break current behavior.

Reply:
I don't think we intend to deprecate the Stackdriver exporter option until we can reach parity with the OpenCensus agent. We can exist in an interim state where the OpenCensus agent works for Prometheus but Stackdriver exporter is still the Stackdriver path.

For monitored resource type support, the correct path forward is to contribute changes in OpenTelemetry Agent which should go Beta soon. Once it goes Beta, OpenTelemetry Agent will be recommended over OpenCensus Agent, and migration should be pretty straightforward when Knative is ready.

Reply:
The OpenCensus API includes a `resource` field which can be set (and we probably should), and seems to be routed to [`ExportMetricsProto`](https://godoc.org/contrib.go.opencensus.io/exporter/stackdriver#Exporter.ExportMetricsProto):

https://github.com/census-instrumentation/opencensus-proto/blob/master/src/opencensus/proto/agent/metrics/v1/metrics_service.proto#L40

Looking at the opencensus code, it looks like the following environment variables may work for setting the resource if you don't have a [`ResourceDectector`](https://godoc.org/contrib.go.opencensus.io/exporter/ocagent#WithResourceDetector) registered: `OC_RESOURCE_TYPE` and `OC_RESOURCE_LABELS`

We could allow setting the ResourceDectector in the metrics opts with a sane non-CloudRun default in `pkg` (which would make sense to me) or set the environment variables appropriately on each component (which seems a bit rougher).